### PR TITLE
A quick and under-featured COM API for ParserState

### LINKED ISSUE #191: Rubberduk.Reflection
Expose `INode` classes to COM, and implement an API that lets VBA code iterate these high-level, wrapped-up parse trees.

### PR DESCRIPTION
**Pull request by retailcoder**
Closes #191 and #1142

### CODE REVIEW

**Comment by ckuhn203** [comment]
Does this close #404 as well? There were several different versions of this request floating around.

**Comment by ThunderFrame** [comment]
What happens if the user is debugging their way through VBA that has calls to the ParserState API, and making changes as they go? Does parsing suspend because it's in Run-Mode?

File: Rubberduck.Parsing/Symbols/Declaration.cs [hunk 1]
```diff
@@ -44,7 +44,7 @@ public class Declaration : IEquatable<Declaration>
             _parentScope = parentScope ?? string.Empty;
             _identifierName = qualifiedName.MemberName;
             _asTypeName = asTypeName;
-            _isSelfAssigned = isSelfAssigned;
+            _isSelfAssigned = isSelfAssigned || (declarationType == DeclarationType.Variable && parentDeclaration != null && parentDeclaration.IdentifierName == ComponentName);
```
Comment (Vogel612) [inline Rubberduck.Parsing/Symbols/Declaration.cs]: Why is determining this the Declaration's responsibility?

I'd have assumed that's what the creator of the object already knows...

File: Rubberduck.Parsing/Symbols/IdentifierReferenceResolver.cs [hunk 2]
```diff
@@ -1067,20 +1073,38 @@ private Declaration FindModuleScopeDeclaration(string identifierName, Declaratio
...
+            var matches = _declarationFinder.MatchName(identifierName).Where(item => !item.Equals(scope)).ToList();
+
             var result = matches.Where(item =>
-                item.ParentScope == localScope.ParentScope
+                (localScope.ParentScopeDeclaration == null || localScope.ParentScopeDeclaration.Equals(item.ParentScopeDeclaration))
                 && !item.DeclarationType.HasFlag(DeclarationType.Member)
                 && !_moduleTypes.Contains(item.DeclarationType)
-                && item.DeclarationType != DeclarationType.UserDefinedType
+                && item.DeclarationType != DeclarationType.UserDefinedType && item.DeclarationType != DeclarationType.Enumeration
                 && (item.DeclarationType != DeclarationType.Event || IsLocalEvent(item, localScope)))
```
Comment (Vogel612) [inline Rubberduck.Parsing/Symbols/IdentifierReferenceResolver.cs]: THis huge predicate could use a name... at least in the diff I can't understand it... maybe the surrounding method gives more information :/

File: Rubberduck.Parsing/Symbols/IdentifierReferenceResolver.cs [hunk 3]
```diff
@@ -1067,20 +1073,38 @@ private Declaration FindModuleScopeDeclaration(string identifierName, Declaratio
...
             .ToList();
 
+            if (matches.Any() && !result.Any())
+            {
+                result = matches.Where(item => 
+                    (localScope != null && localScope.Equals(item.ParentScopeDeclaration))
+                    && !item.DeclarationType.HasFlag(DeclarationType.Member)
+                    && !_moduleTypes.Contains(item.DeclarationType)
+                    && item.DeclarationType != DeclarationType.UserDefinedType && item.DeclarationType != DeclarationType.Enumeration
+                    && (item.DeclarationType != DeclarationType.Event || IsLocalEvent(item, localScope)))
```
Comment (Vogel612) [inline Rubberduck.Parsing/Symbols/IdentifierReferenceResolver.cs]: Save the first line, this is basically the same predicate as above ... what happened to DRY? :)

Comment (retailcoder) [inline Rubberduck.Parsing/Symbols/IdentifierReferenceResolver.cs]: IKR ...desperate measures... I had been playing whack-a-mole with resolver tests for over two hours - now that they all pass, I'm fine with a refactoring of the resolver code. This is code from someone that desperately wants to move on to other parts of the duck.

**Comment by retailcoder** [comment]
@ckuhn203 I considered closing #404 with it, but the API is a little too bare-bones for now. It needs to expose parser errors and some helper methods to retrieve e.g. all member procedures, among other things.

@ThunderFrame if the user makes changes, they make their parser state stale, simple as that - the `ParserState` instance lives in the VBA code and does not interefere with Rubberduck's own parser state. I made the `Parse` method a blocking call on purpose, but the intent is to also use `BeginParse` to hook up parser state events (and handle them with a `WithEvents` variable). Not sure I want to expose component-level parsing at that point though, not until the resolver can handle "surgical" resolving of references (i.e. v3.0).

**Pull request merged**